File: src/main/java/com/sems/analytics/interfaces/rest/resources/AnalyticsResources.java
package com.sems.analytics.interfaces.rest.resources;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import com.sems.analytics.domain.model.entities.*;
import com.sems.analytics.domain.model.valueobjects.RankingItem;
import jakarta.validation.constraints.NotBlank;
import java.time.Instant;
import java.util.List;

/**
 * Contrato JSON del modulo de analitica.
 *
 * <p>Como el resto de lo que venia de FastAPI, se serializa en snake_case.
 */
public final class AnalyticsResources {

    private AnalyticsResources() {
    }

    // ------------------------------------------------------------- peticiones

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record CreatePredictionRequest(
            @NotBlank(message = "is required") String userId,
            int predictionYear, int predictionMonth,
            Instant periodStart, Instant periodEnd,
            double estimatedKwh, double estimatedAmount,
            String currency, double tariffUsed, double errorMarginPercentage) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record CreateRecommendationRequest(
            @NotBlank(message = "is required") String userId,
            String deviceId, String recommendationType, String title, String description,
            double estimatedSavingKwh, double estimatedSavingAmount, String currency) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record CreateAnomalyRequest(
            @NotBlank(message = "is required") String userId,
            String deviceId, String anomalyType, String description, String severity,
            double actualKwh, double expectedKwh) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record CreateIdentificationRequest(
            @NotBlank(message = "is required") String userId,
            String deviceId, String predictedDeviceType, double confidenceScore, String status) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record RankingItemResource(int rank, String deviceId, String deviceName, double totalKwh,
                                      double estimatedAmount, double percentageOfTotal, String currency) {

        public static RankingItemResource from(RankingItem i) {
            return new RankingItemResource(i.rank(), i.deviceId(), i.deviceName(), i.totalKwh(),
                    i.estimatedAmount(), i.percentageOfTotal(), i.currency());
        }

        public RankingItem toDomain() {
            return new RankingItem(rank, deviceId, deviceName, totalKwh, estimatedAmount,
                    percentageOfTotal, currency);
        }
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record CreateRankingRequest(
            @NotBlank(message = "is required") String userId,
            String periodType, Instant periodStart, Instant periodEnd,
            List<RankingItemResource> rankings) {
    }

    // -------------------------------------------------------------- respuestas

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record BillPredictionResponse(
            String id, String userId, int predictionYear, int predictionMonth,
            Instant periodStart, Instant periodEnd, double estimatedKwh, double estimatedAmount,
            String currency, double tariffUsed, double errorMarginPercentage,
            Instant generatedAt, Instant createdAt) {

        public static BillPredictionResponse from(BillPrediction p) {
            return new BillPredictionResponse(p.getId().toString(), p.getUserId(),
                    p.getPredictionYear(), p.getPredictionMonth(), p.getPeriodStart(),
                    p.getPeriodEnd(), p.getEstimatedKwh(), p.getEstimatedAmount(), p.getCurrency(),
                    p.getTariffUsed(), p.getErrorMarginPercentage(), p.getGeneratedAt(),
                    p.getCreatedAt());
        }
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record RecommendationResponse(
            String id, String userId, String deviceId, String recommendationType, String title,
            String description, double estimatedSavingKwh, double estimatedSavingAmount,
            String currency, String status, Instant generatedAt, Instant appliedAt, Instant createdAt) {

        public static RecommendationResponse from(Recommendation r) {
            return new RecommendationResponse(r.getId().toString(), r.getUserId(), r.getDeviceId(),
                    r.getRecommendationType(), r.getTitle(), r.getDescription(),
                    r.getEstimatedSavingKwh(), r.getEstimatedSavingAmount(), r.getCurrency(),
                    r.getStatus(), r.getGeneratedAt(), r.getAppliedAt(), r.getCreatedAt());
        }
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record AnomalyResponse(
            String id, String userId, String deviceId, String anomalyType, String description,
            String severity, String status, double actualKwh, double expectedKwh,
            double deviationPercentage, Instant detectedAt, Instant resolvedAt, Instant createdAt) {

        public static AnomalyResponse from(Anomaly a) {
            return new AnomalyResponse(a.getId().toString(), a.getUserId(), a.getDeviceId(),
                    a.getAnomalyType(), a.getDescription(), a.getSeverity(), a.getStatus(),
                    a.getActualKwh(), a.getExpectedKwh(), a.getDeviationPercentage(),
                    a.getDetectedAt(), a.getResolvedAt(), a.getCreatedAt());
        }
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record DeviceIdentificationResponse(
            String id, String userId, String deviceId, String predictedDeviceType,
            double confidenceScore, String status, Instant analyzedAt, Instant createdAt) {

        public static DeviceIdentificationResponse from(DeviceIdentificationResult d) {
            return new DeviceIdentificationResponse(d.getId().toString(), d.getUserId(),
                    d.getDeviceId(), d.getPredictedDeviceType(), d.getConfidenceScore(),
                    d.getStatus(), d.getAnalyzedAt(), d.getCreatedAt());
        }
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record ConsumptionRankingResponse(
            String id, String userId, String periodType, Instant periodStart, Instant periodEnd,
            List<RankingItemResource> rankings, Instant generatedAt, Instant createdAt) {

        public static ConsumptionRankingResponse from(ConsumptionRanking c) {
            return new ConsumptionRankingResponse(c.getId().toString(), c.getUserId(),
                    c.getPeriodType(), c.getPeriodStart(), c.getPeriodEnd(),
                    c.getRankings().stream().map(RankingItemResource::from).toList(),
                    c.getGeneratedAt(), c.getCreatedAt());
        }
    }
}

File: src/Sems.Api/Modules/Analytics/Interfaces/AnalyticsResources.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;
using Sems.Api.Modules.Analytics.Domain.Model;

namespace Sems.Api.Modules.Analytics.Interfaces;

/// <summary>
/// Contrato JSON del modulo de analitica.
///
/// <para>Como el resto de lo que venia de FastAPI, se serializa en snake_case y
/// cada campo lo declara explicitamente.</para>
/// </summary>
public static class AnalyticsResources
{
    // ------------------------------------------------------------- peticiones

    public sealed record CreatePredictionRequest(
        [property: JsonPropertyName("user_id")]
        [property: Required(ErrorMessage = "is required")] string UserId,
        [property: JsonPropertyName("prediction_year")] int PredictionYear,
        [property: JsonPropertyName("prediction_month")] int PredictionMonth,
        [property: JsonPropertyName("period_start")] DateTime PeriodStart,
        [property: JsonPropertyName("period_end")] DateTime PeriodEnd,
        [property: JsonPropertyName("estimated_kwh")] double EstimatedKwh,
        [property: JsonPropertyName("estimated_amount")] double EstimatedAmount,
        [property: JsonPropertyName("currency")] string? Currency,
        [property: JsonPropertyName("tariff_used")] double TariffUsed,
        [property: JsonPropertyName("error_margin_percentage")] double ErrorMarginPercentage);

    public sealed record CreateRecommendationRequest(
        [property: JsonPropertyName("user_id")]
        [property: Required(ErrorMessage = "is required")] string UserId,
        [property: JsonPropertyName("device_id")] string? DeviceId,
        [property: JsonPropertyName("recommendation_type")] string? RecommendationType,
        [property: JsonPropertyName("title")] string? Title,
        [property: JsonPropertyName("description")] string? Description,
        [property: JsonPropertyName("estimated_saving_kwh")] double EstimatedSavingKwh,
        [property: JsonPropertyName("estimated_saving_amount")] double EstimatedSavingAmount,
        [property: JsonPropertyName("currency")] string? Currency);

    public sealed record CreateAnomalyRequest(
        [property: JsonPropertyName("user_id")]
        [property: Required(ErrorMessage = "is required")] string UserId,
        [property: JsonPropertyName("device_id")] string? DeviceId,
        [property: JsonPropertyName("anomaly_type")] string? AnomalyType,
        [property: JsonPropertyName("description")] string? Description,
        [property: JsonPropertyName("severity")] string? Severity,
        [property: JsonPropertyName("actual_kwh")] double ActualKwh,
        [property: JsonPropertyName("expected_kwh")] double ExpectedKwh);

    public sealed record CreateIdentificationRequest(
        [property: JsonPropertyName("user_id")]
        [property: Required(ErrorMessage = "is required")] string UserId,
        [property: JsonPropertyName("device_id")] string? DeviceId,
        [property: JsonPropertyName("predicted_device_type")] string? PredictedDeviceType,
        [property: JsonPropertyName("confidence_score")] double ConfidenceScore,
        [property: JsonPropertyName("status")] string? Status);

    public sealed record RankingItemResource(
        [property: JsonPropertyName("rank")] int Rank,
        [property: JsonPropertyName("device_id")] string DeviceId,
        [property: JsonPropertyName("device_name")] string DeviceName,
        [property: JsonPropertyName("total_kwh")] double TotalKwh,
        [property: JsonPropertyName("estimated_amount")] double EstimatedAmount,
        [property: JsonPropertyName("percentage_of_total")] double PercentageOfTotal,
        [property: JsonPropertyName("currency")] string Currency)
    {
        public static RankingItemResource From(RankingItem i) => new(i.Rank, i.DeviceId,
            i.DeviceName, i.TotalKwh, i.EstimatedAmount, i.PercentageOfTotal, i.Currency);

        public RankingItem ToDomain() => new(Rank, DeviceId, DeviceName, TotalKwh, EstimatedAmount,
            PercentageOfTotal, Currency);
    }

    public sealed record CreateRankingRequest(
        [property: JsonPropertyName("user_id")]
        [property: Required(ErrorMessage = "is required")] string UserId,
        [property: JsonPropertyName("period_type")] string? PeriodType,
        [property: JsonPropertyName("period_start")] DateTime PeriodStart,
        [property: JsonPropertyName("period_end")] DateTime PeriodEnd,
        [property: JsonPropertyName("rankings")] List<RankingItemResource>? Rankings);

    // -------------------------------------------------------------- respuestas

    public sealed record BillPredictionResponse(
        [property: JsonPropertyName("id")] string Id,
        [property: JsonPropertyName("user_id")] string UserId,
        [property: JsonPropertyName("prediction_year")] int PredictionYear,
        [property: JsonPropertyName("prediction_month")] int PredictionMonth,
        [property: JsonPropertyName("period_start")] DateTime PeriodStart,
        [property: JsonPropertyName("period_end")] DateTime PeriodEnd,
        [property: JsonPropertyName("estimated_kwh")] double EstimatedKwh,
        [property: JsonPropertyName("estimated_amount")] double EstimatedAmount,
        [property: JsonPropertyName("currency")] string Currency,
        [property: JsonPropertyName("tariff_used")] double TariffUsed,
        [property: JsonPropertyName("error_margin_percentage")] double ErrorMarginPercentage,
        [property: JsonPropertyName("generated_at")] DateTime GeneratedAt,
        [property: JsonPropertyName("created_at")] DateTime CreatedAt)
    {
        public static BillPredictionResponse From(BillPrediction p) => new(p.Id.ToString(),
            p.UserId, p.PredictionYear, p.PredictionMonth, p.PeriodStart, p.PeriodEnd,
            p.EstimatedKwh, p.EstimatedAmount, p.Currency, p.TariffUsed, p.ErrorMarginPercentage,
            p.GeneratedAt, p.CreatedAt);
    }

    public sealed record RecommendationResponse(
        [property: JsonPropertyName("id")] string Id,
        [property: JsonPropertyName("user_id")] string UserId,
        [property: JsonPropertyName("device_id")] string? DeviceId,
        [property: JsonPropertyName("recommendation_type")] string? RecommendationType,
        [property: JsonPropertyName("title")] string? Title,
        [property: JsonPropertyName("description")] string? Description,
        [property: JsonPropertyName("estimated_saving_kwh")] double EstimatedSavingKwh,
        [property: JsonPropertyName("estimated_saving_amount")] double EstimatedSavingAmount,
        [property: JsonPropertyName("currency")] string Currency,
        [property: JsonPropertyName("status")] string Status,
        [property: JsonPropertyName("generated_at")] DateTime GeneratedAt,
        [property: JsonPropertyName("applied_at")] DateTime? AppliedAt,
        [property: JsonPropertyName("created_at")] DateTime CreatedAt)
    {
        public static RecommendationResponse From(Recommendation r) => new(r.Id.ToString(),
            r.UserId, r.DeviceId, r.RecommendationType, r.Title, r.Description,
            r.EstimatedSavingKwh, r.EstimatedSavingAmount, r.Currency, r.Status, r.GeneratedAt,
            r.AppliedAt, r.CreatedAt);
    }

    public sealed record AnomalyResponse(
        [property: JsonPropertyName("id")] string Id,
        [property: JsonPropertyName("user_id")] string UserId,
        [property: JsonPropertyName("device_id")] string? DeviceId,
        [property: JsonPropertyName("anomaly_type")] string? AnomalyType,
        [property: JsonPropertyName("description")] string? Description,
        [property: JsonPropertyName("severity")] string? Severity,
        [property: JsonPropertyName("status")] string Status,
        [property: JsonPropertyName("actual_kwh")] double ActualKwh,
        [property: JsonPropertyName("expected_kwh")] double ExpectedKwh,
        [property: JsonPropertyName("deviation_percentage")] double DeviationPercentage,
        [property: JsonPropertyName("detected_at")] DateTime DetectedAt,
        [property: JsonPropertyName("resolved_at")] DateTime? ResolvedAt,
        [property: JsonPropertyName("created_at")] DateTime CreatedAt)
    {
        public static AnomalyResponse From(Anomaly a) => new(a.Id.ToString(), a.UserId, a.DeviceId,
            a.AnomalyType, a.Description, a.Severity, a.Status, a.ActualKwh, a.ExpectedKwh,
            a.DeviationPercentage, a.DetectedAt, a.ResolvedAt, a.CreatedAt);
    }

    public sealed record DeviceIdentificationResponse(
        [property: JsonPropertyName("id")] string Id,
        [property: JsonPropertyName("user_id")] string UserId,
        [property: JsonPropertyName("device_id")] string? DeviceId,
        [property: JsonPropertyName("predicted_device_type")] string? PredictedDeviceType,
        [property: JsonPropertyName("confidence_score")] double ConfidenceScore,
        [property: JsonPropertyName("status")] string Status,
        [property: JsonPropertyName("analyzed_at")] DateTime AnalyzedAt,
        [property: JsonPropertyName("created_at")] DateTime CreatedAt)
    {
        public static DeviceIdentificationResponse From(DeviceIdentificationResult d) =>
            new(d.Id.ToString(), d.UserId, d.DeviceId, d.PredictedDeviceType, d.ConfidenceScore,
                d.Status, d.AnalyzedAt, d.CreatedAt);
    }

    public sealed record ConsumptionRankingResponse(
        [property: JsonPropertyName("id")] string Id,
        [property: JsonPropertyName("user_id")] string UserId,
        [property: JsonPropertyName("period_type")] string? PeriodType,
        [property: JsonPropertyName("period_start")] DateTime PeriodStart,
        [property: JsonPropertyName("period_end")] DateTime PeriodEnd,
        [property: JsonPropertyName("rankings")] List<RankingItemResource> Rankings,
        [property: JsonPropertyName("generated_at")] DateTime GeneratedAt,
        [property: JsonPropertyName("created_at")] DateTime CreatedAt)
    {
        public static ConsumptionRankingResponse From(ConsumptionRanking c) => new(c.Id.ToString(),
            c.UserId, c.PeriodType, c.PeriodStart, c.PeriodEnd,
            c.Rankings().Select(RankingItemResource.From).ToList(), c.GeneratedAt, c.CreatedAt);
    }
}
